plugins/ranking_procedure.cc

```diff
@@ -0,0 +1,181 @@
+}
+
+RankingProcedureConfig::
+RankingProcedureConfig()
+{
+    outputDataset.withType("sparse.mutable");
+}
+
+DEFINE_STRUCTURE_DESCRIPTION(RankingProcedureConfig);
+
+RankingProcedureConfigDescription::
+RankingProcedureConfigDescription()
+{
+    addField("inputData", &RankingProcedureConfig::inputData,
+             "An SQL statement to select the input data. The select "
+             "expression is required but has no effect. The order by "
+             "expression is used to rank the rows.");
+    addField("outputDataset", &RankingProcedureConfig::outputDataset,
+             "Output dataset configuration. This may refer either to an "
+             "existing dataset, or a fully specified but non-existing dataset "
+             "which will be created by the procedure.",
+             PolyConfigT<Dataset>().withType("sparse.mutable"));
+    addField("rankingType", &RankingProcedureConfig::rankingType,
+             "The type of the rank to output. Either percentile or index");
+    addParent<ProcedureConfig>();
+}
+
+RankingProcedure::
+RankingProcedure(MldbServer * owner,
+                 PolyConfig config,
+                 const std::function<bool (const Json::Value &)> & onProgress)
+    : Procedure(owner)
+{
+    procedureConfig = config.params.convert<RankingProcedureConfig>();
+}
+
+RunOutput
+RankingProcedure::
+run(const ProcedureRunConfig & run,
+    const std::function<bool (const Json::Value &)> & onProgress) const
+{
+    SqlExpressionMldbContext context(server);
+
+    auto boundDataset = procedureConfig.inputData.stm->from->bind(context);
+
+    SelectExpression select(SelectExpression::parse("1"));
+    vector<shared_ptr<SqlExpression> > calc;
+
+    vector<Id> orderedRowNames;
+    auto getSize = [&] (const MatrixNamedRow & row,
+                        const vector<ExpressionValue> & calc)
+    {
+        orderedRowNames.emplace_back(row.rowName);
+        return true;
+    };
+
+    BoundSelectQuery(select,
+                     *boundDataset.dataset,
+                     boundDataset.asName,
+                     procedureConfig.inputData.stm->when,
+                     procedureConfig.inputData.stm->where,
+                     procedureConfig.inputData.stm->orderBy,
+                     calc)
+        .execute(getSize,
+                 procedureConfig.inputData.stm->offset,
+                 procedureConfig.inputData.stm->limit,
+                 onProgress);
+
+    int64_t rowCount = orderedRowNames.size();
+
+    auto output = createDataset(server, procedureConfig.outputDataset,
+                                nullptr, true /*overwrite*/);
+
+    typedef tuple<ColumnName, CellValue, Date> cell;
+    PerThreadAccumulator<vector<pair<RowName, vector<cell>>>> accum;
+    const ColumnName columnName("rank");
+    function<void(int64_t)> applyFct;
+    float countD100 = (rowCount - 1) / 100.0;
+    if (procedureConfig.rankingType == RankingType::PERCENTILE) {
+        applyFct = [&](int64_t idx)
```

Comment:
You could parametrize the callback with the RankingType and bind it at execution to avoid the code duplication.